apps/emqx_durable_storage/src/emqx_ds_replication_layer.erl

```diff
@@ -344,3 +339,167 @@ shard_of_messages(DB, [#message{from = From} | _]) ->
+    {_Name, Node} = ra_random_replica(DB, Shard),
+    emqx_ds_proto_v2:get_streams(Node, DB, Shard, TopicFilter, Time).
+
+ra_make_iterator(DB, Shard, Stream, TopicFilter, StartTime) ->
+    {_Name, Node} = ra_random_replica(DB, Shard),
+    emqx_ds_proto_v2:make_iterator(Node, DB, Shard, Stream, TopicFilter, StartTime).
+
+ra_update_iterator(DB, Shard, Iter, DSKey) ->
+    {_Name, Node} = ra_random_replica(DB, Shard),
+    emqx_ds_proto_v2:update_iterator(Node, DB, Shard, Iter, DSKey).
+
+ra_next(DB, Shard, Iter, BatchSize) ->
+    {_Name, Node} = ra_random_replica(DB, Shard),
+    emqx_ds_proto_v2:next(Node, DB, Shard, Iter, BatchSize).
+
+ra_drop_shard(DB, Shard) ->
+    %% TODO: clean dsrepl state
+    ra:stop_server(_System = default, ra_local_server(DB, Shard)).
+
+ra_shard_servers(DB, Shard) ->
+    {ok, ReplicaSet} = emqx_ds_replication_layer_meta:replica_set(DB, Shard),
+    [
+        {ra_server_name(DB, Shard, Site), emqx_ds_replication_layer_meta:node(Site)}
+     || Site <- ReplicaSet
+    ].
+
+ra_local_server(DB, Shard) ->
+    Site = emqx_ds_replication_layer_meta:this_site(),
+    {ra_server_name(DB, Shard, Site), node()}.
+
+ra_leader_servers(DB, Shard) ->
+    %% NOTE: Contact last known leader first, then rest of shard servers.
+    ClusterName = ra_cluster_name(DB, Shard),
+    case ra_leaderboard:lookup_leader(ClusterName) of
+        Leader when Leader /= undefined ->
+            Servers = ra_leaderboard:lookup_members(ClusterName),
+            [Leader | lists:delete(Leader, Servers)];
+        undefined ->
+            %% TODO: Dynamic membership.
+            ra_shard_servers(DB, Shard)
+    end.
+
+ra_random_replica(DB, Shard) ->
+    %% NOTE: Contact random replica that is not a known leader.
+    %% TODO: Replica may be down, so we may need to retry.
+    ClusterName = ra_cluster_name(DB, Shard),
+    case ra_leaderboard:lookup_members(ClusterName) of
+        Servers when is_list(Servers) ->
+            Leader = ra_leaderboard:lookup_leader(ClusterName),
+            ra_pick_replica(Servers, Leader);
+        undefined ->
+            %% TODO
+            %% Leader is unkonwn if there are no servers of this group on the
+            %% local node. We want to pick a replica in that case as well.
+            %% TODO: Dynamic membership.
+            ra_pick_server(ra_shard_servers(DB, Shard))
+    end.
+
+ra_pick_replica(Servers, Leader) ->
+    case lists:delete(Leader, Servers) of
+        [] ->
+            Leader;
+        Followers ->
+            ra_pick_server(Followers)
+    end.
+
+ra_pick_server(Servers) ->
+    lists:nth(rand:uniform(length(Servers)), Servers).
+
+ra_cluster_name(DB, Shard) ->
+    iolist_to_binary(io_lib:format("~s_~s", [DB, Shard])).
+
+ra_server_name(DB, Shard, Site) ->
+    DBBin = atom_to_binary(DB),
+    binary_to_atom(<<"ds_", DBBin/binary, Shard/binary, "_", Site/binary>>).
+
+%%
+
+init(#{db := DB, shard := Shard}) ->
+    _ = erlang:put(emqx_ds_db_shard, {DB, Shard}),
```

Comment:
Q: is there a particular reason not to put this in the state map?

Reply:
My line of thinking is that `State` is essentially a product of applying Raft log entries, and is periodically snapshotted and dumped to disk. This tuple feels odd as part of state / snapshot, it's more like a runtime-only information needed to locate relevant Erlang processes. On the other hand I agree that it feels hacky to put it in the process dictionary.

Reply:
I wouldn't object to using process dictionary too much, as long as the terms are static. 
Process dictionary becomes rather sticky when it's used to store temporary data, because it doesn't get cleaned up on exception.